pkg/services/baremetal/baremetal/baremetal.go

```diff
@@ -363,71 +365,126 @@ func (s *Service) getAssociatedHost(ctx context.Context) (*infrav1.HetznerBareMe
 	return &host, helper, nil
 }
 
-func (s *Service) chooseHost(ctx context.Context) (*infrav1.HetznerBareMetalHost, *patch.Helper, error) {
+// chooseHost tries to find a free hbmh.
+// If no hbmh was found, then hbmh and err are nil, and the string
+// "reason" contains human readable details.
+func (s *Service) chooseHost(ctx context.Context) (
+	hbmh *infrav1.HetznerBareMetalHost, patchHelper *patch.Helper, reason string, err error,
+) {
 	// get list of hosts scoped to namespace of machine
 	hosts := infrav1.HetznerBareMetalHostList{}
 	opts := &client.ListOptions{
 		Namespace: s.scope.BareMetalMachine.Namespace,
 	}
 
 	if err := s.scope.Client.List(ctx, &hosts, opts); err != nil {
-		return nil, nil, fmt.Errorf("failed to list hosts: %w", err)
+		return nil, nil, "", fmt.Errorf("failed to list hosts: %w", err)
 	}
 
 	labelSelector := s.getLabelSelector()
 
-	availableHosts := make([]*infrav1.HetznerBareMetalHost, 0, len(hosts.Items))
+	usableHosts := make([]*infrav1.HetznerBareMetalHost, 0, len(hosts.Items))
 
+	mapOfSkipReasons := make(map[string]int)
+
+	availableDevices := 0
 	for i, host := range hosts.Items {
 		if host.Spec.ConsumerRef != nil && consumerRefMatches(host.Spec.ConsumerRef, s.scope.BareMetalMachine) {
 			helper, err := patch.NewHelper(&hosts.Items[i], s.scope.Client)
 			if err != nil {
-				return nil, nil, fmt.Errorf("failed to create patch helper: %w", err)
+				return nil, nil, "", fmt.Errorf("failed to create patch helper: %w", err)
 			}
-			return &hosts.Items[i], helper, nil
+			return &hosts.Items[i], helper, "", nil
 		}
 		if host.Spec.ConsumerRef != nil {
 			continue
 		}
-		if host.Spec.MaintenanceMode != nil && *host.Spec.MaintenanceMode {
+		if host.Spec.Status.ProvisioningState != infrav1.StateNone {
 			continue
 		}
+
+		// from now on each "continue" should add an entry
+		// to mapOfSkipReasons.
+		availableDevices++
+
+		// This comes first, because we should not look too deep into machines
+		// which are not in our scope.
+		if !labelSelector.Matches(labels.Set(host.ObjectMeta.Labels)) {
+			mapOfSkipReasons["label-selector-does-not-match"]++
+			continue
+		}
+
 		if host.GetDeletionTimestamp() != nil {
+			mapOfSkipReasons["hbmh-has-deletion-timestamp"]++
+			continue
+		}
+		if host.Spec.MaintenanceMode != nil && *host.Spec.MaintenanceMode {
+			mapOfSkipReasons["hbmh-in-maintenance-mode"]++
 			continue
 		}
 		if host.Spec.Status.ErrorMessage != "" {
+			mapOfSkipReasons["hbmh-has-error-message-in-status"]++
 			continue
 		}
 
-		if !labelSelector.Matches(labels.Set(host.ObjectMeta.Labels)) {
-			continue
+		// This comes last, because now we would choose the machine, but we check
+		// if the config is correct.
+		if host.Spec.RootDeviceHints != nil &&
+			host.Spec.RootDeviceHints.IsValid() {
+			// If there are no RootDeviceHints, then the it should be ok that the hbmh gets picked.
```

Comment:
better text:
Even if there are no rootDeviceHints specified, the host is usable. After the phase registering, the process to provision the server stops and waits for the user to specify the rootDeviceHints. 
Here we want to check whether the specified rootDeviceHints fit with the InstallImage configuration of the HetznerBareMetalMachine. If not, it is not usable.

Reply:
yes, makes sense. Does the text of 1106b11b fit?